cluster mempool: control/optimize TxGraph memory usage

sipa opened the pull request:
Part of #30289.

This adds a few optimizations to reduce `TxGraph`'s memory usage, and makes sure that dynamic memory it uses doesn't linger after shrinking clusters. Finally, it exposes a function `GetMainMemoryUsage()` to compute `TxGraph`'s approximate memory usage.

It makes the `Cluster` type abstract, with two instances (`SingletonClusterImpl` for 1-transaction clusters, and `GenericClusterImpl` for others).

On my 64-bit system, I obtain the following numbers:
* `SingletonClusterImpl`: 48 bytes, plus 16 bytes malloc overhead in its `unique_ptr`, plus 8-byte pointer in `m_clusters`
* `GenericClusterImpl`: 104 bytes, plus 16 bytes malloc overhead in its `unique_ptr`, plus 8-byte pointer in `m_clusters`, plus 72 bytes malloc overhead inside its vectors and `DepGraph`, plus 40 bytes per transaction in those.
* `TxGraphImpl::Entry`: 72 bytes per transaction
* `TxGraphImpl::ChunkData`: 8 bytes, plus 56 bytes in `std::set` overhead + malloc overhead, all per chunk.
* `TxGraph::Ref`: 16 bytes per transaction

This overall amounts to 200 bytes per cluster, plus 64 bytes per chunk, plus 128 bytes per transaction, but only 224 bytes overall per singleton cluster.

sipa committed:
txgraph: Make level of Cluster implicit (optimization)

This reduces per-Cluster memory usage by making Clusters not aware of their
own level. Instead, track it either in calling code, or infer it based on
the transactions in them.

sipa committed:
txgraph: avoid holes in DepGraph positions (mem optimization)

sipa committed:
depgraph: add memory usage control (feature)

Co-Authored-By: Lőrinc <pap.lorinc@gmail.com>

sipa committed:
txgraph: keep data structures compact (mem optimization)

sipa committed:
txgraph: keep track of Cluster memory usage (preparation)

sipa committed:
txgraph: expose memory usage estimate function (feature)

ismaelsadeeq reviewed (commented):
Concept ACK to have control over memory usage, also not suprised by the numbers because of the added complexity of cluster mempool v master.

instagibbs reviewed (commented):
concept ACK, I refuse to do math

sipa committed:
txgraph: avoid accessing other Cluster internals (refactor)

This adds 4 functions to Cluster to help implement Merge() and Split() without
needing access to the internals of the other Cluster. This is a preparation for
a follow-up that will make Clusters a virtual class whose internals are abstracted
away.

sipa committed:
txgraph: add SingletonClusterImpl (mem optimization)

This adds a specialized Cluster implementation for singleton clusters, saving
a significant amount of memory by avoiding the need for m_depgraph, m_mapping,
and m_linearization, and their overheads.

sipa committed:
txgraph: abstract out creation of empty Clusters (refactor)

sipa committed:
txgraph: move some sanity checks from Cluster to TxGraphImpl (refactor)

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -161,10 +159,13 @@ class Cluster
     void UpdateMapping(DepGraphIndex cluster_idx, GraphIndex graph_idx) noexcept { m_mapping[cluster_idx] = graph_idx; }
     /** Push changes to Cluster and its linearization to the TxGraphImpl Entry objects. */
-    void Updated(TxGraphImpl& graph) noexcept;
+    void Updated(TxGraphImpl& graph, int level) noexcept;
```
Thanks for the context.
For anyone that wants to dig in, the `testmempoolaccept` third level was discussed here https://github.com/bitcoin/bitcoin/issues/32160

ismaelsadeeq reviewed (commented):
Code Review ACK 66401148ddcba04c27751e39e53fc41e6deb5d14

I shutdown a long-running node at height 915249.

```
.rw------- 128M ismaelsadeeq 18 Sep 14:46 mempool.dat
```

Restarted the node on master 1444ed855f4 with:

  ```bash
  build/bin/bitcoind -connect=0 -debug
  ```

After loading the mempool, `getmempoolinfo` returned:

  ```json
  {
    "loaded": true,
    "size": 104944,
    "bytes": 43254588,
    "usage": 260130384,
    "total_fee": 0.10523099,
    "maxmempool": 300000000
  }
  ```

I restarted the node on full cluster mempool implementation 57f929d with the same config.
  `getmempoolinfo` returned:

  ```json
  {
    "loaded": true,
    "size": 104944,
    "bytes": 43254588,
    "usage": 244614928,
    "total_fee": 0.10523099,
    "maxmempool": 300000000
  }
  ```

The mempool size and fee data are identical across runs, but `usage` differ with cluster mempool having slightly less usage at  6%.

instagibbs reviewed (approved):
ACK 66401148ddcba04c27751e39e53fc41e6deb5d14

Did not validate the numbers, non-blocking comments only

ismaelsadeeq commented:
reACK 5ac32aa441bf3ec1f1fa826f689405bc4a7b7354

Changes since last ACK are nits fixes.

l0rinc reviewed (commented):
It's my first time diving into cluster mempool, and it seems like a solid implementation.
I like the polymorphic split between `SingletonClusterImpl` and `GenericClusterImpl`.

I've documented my review journey and left many comments. Some are tiny nits, but others appear to be leftovers from previous refactorings that should be corrected - I'm fine with doing them in a follow-up PR as well.
If you decide to apply any suggestions here, I've included a few commit reorganization tips that could make the changes clearer.

Key suggestions include:
- Fuzz test enhancements for more thorough coverage
- Documentation fixes for stale references from previous refactorings  
- Dead code removal after recent memory compaction changes
- Comments highlighting surprising side-effects in certain methods

I don't yet feel confident enough to fully approve, but this is definitely a strong Concept ACK for the change.

glozow reviewed (commented):
code review ACK 5ac32aa441b 

I focused on checking that all the dynamic memory is accounted for, places where `m_cluster_usage` is updated, and correctness of the `SingletonCluster` implementation. Feel free to ignore my nits.

l0rinc commented on src/txgraph.cpp:
```diff
@@ -227,8 +238,8 @@ class Cluster
 class GenericClusterImpl final : public Cluster
 {
     friend class TxGraphImpl;
+    using SetType = Cluster::SetType;
```
We're in `Cluser` so `SetType` should be available here, so I think it's redundant

l0rinc commented on src/txgraph.cpp:
```diff
@@ -1071,6 +1099,8 @@ void Cluster::Merge(TxGraphImpl& graph, int level, Cluster& other) noexcept
 {
     /** Vector to store the positions in this Cluster for each position in other. */
     std::vector<DepGraphIndex> remap(other.m_depgraph.PositionRange());
+    graph.GetClusterSet(level).m_cluster_usage -= TotalMemoryUsage();
+    graph.GetClusterSet(level).m_cluster_usage -= other.TotalMemoryUsage();
```
I'm also fine with other ways of deduplication - or to leave as is.

sipa committed:
txgraph: make Cluster an abstract class (refactor)

sipa committed:
txgraph: give Clusters a range of intended tx counts (preparation)

sipa committed:
txgraph: comment fixes (doc fix)

instagibbs reviewed (approved):
reACK 7421e58dee402ee8b5b58a18684953a89ad350d7

`git range-diff master 5ac32aa441bf3ec1f1fa826f689405bc4a7b7354 7421e58dee402ee8b5b58a18684953a89ad350d7`

ismaelsadeeq reviewed (commented):
reACK 023cd5a5469ad61205bf7bb1135895f2b4a20ea9  🚢  

[ad350d7..023cd5a](https://github.com/bitcoin/bitcoin/compare/7421e58dee402ee8b5b58a18684953a89ad350d7..023cd5a5469ad61205bf7bb1135895f2b4a20ea9)

Several nits from recent comments were addressed, clarifying comments were added for clarity.

glozow reviewed (commented):
reACK 023cd5a5469ad61205bf7bb1135895f2b4a20ea9

Thanks for taking the suggestions!

l0rinc commented on src/cluster_linearize.h:
pushed a follow-up to https://github.com/bitcoin/bitcoin/pull/33768
